hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -688,6 +663,48 @@ public OmKeyInfo lookupKey(OmKeyArgs args, String clientAddress)
     }
   }
 
+  /**
+   * Refresh pipeline info in OM by asking SCM.
+   * @param value OmKeyInfo
+   */
+  @VisibleForTesting
+  protected void refreshPipeline(OmKeyInfo value) {
+    // Refresh container pipeline info from SCM
+    // based on OmKeyArgs.refreshPipeline flag
+    // 1. Client send initial read request OmKeyArgs.refreshPipeline = false
+    // and uses the pipeline cached in OM to access datanode
+    // 2. If succeeded, done.
+    // 3. If failed due to pipeline does not exist or invalid pipeline state
+    //    exception, client should retry lookupKey with
+    //    OmKeyArgs.refreshPipeline = true
+    Map<Long, ContainerWithPipeline> containerWithPipelineMap = new HashMap<>();
+    for (OmKeyLocationInfoGroup key : value.getKeyLocationVersions()) {
+      key.getLocationList().forEach(k -> {
+        // TODO: fix Some tests that may not initialize container client
+        // The production should always have containerClient initialized.
+        if (scmClient.getContainerClient() != null) {
+          ContainerWithPipeline cp = containerWithPipelineMap
+              .computeIfAbsent(k.getContainerID(), cId -> {
+                try {
+                  return scmClient.getContainerClient()
+                      .getContainerWithPipeline(cId);
+                } catch (IOException e) {
+                  LOG.error("Unable to update pipeline for container:{}",
+                      k.getContainerID());
+                  return null;
+                }
+              });
+          if (cp == null) {
+            return;
```

Comment:
Do we need to fail here? As for that block, We are not finding ContainerWithPipeline.

Reply:
The old logic would have thrown an NPE in Line 665 if the ContainerWithPipeline got from SCM is null. I have changed that to a return from method here. I have not added a fail fast logic.

Reply:
So, in this case now with current code client will eventually fail when reading that container?
Why can't we fail here itself, any advantage we get with delaying? (As now in Client, we don't have logic for retry with refresh=true if it failed to ready the block/container, I believe)

Reply:
Refactored the logic to throw exception when the call to SCM fails.